File: scripts/record_video_server.py
#!/usr/bin/env python
import rospy
import cv2
import cv_bridge
from sensor_msgs.msg import Image
from streaming_media.srv import *
from std_msgs.msg import Bool
from threading import Lock

recording = False
isTimeOut = False
base_save_video_path = "/home/videos/"
framesOfRGB = []
framesOfDepth = []

def video_capture_callback_rgb(msg):
    global recording, framesOfRGB, framesOfDepth
    if not recording:
        return
    try:
        cv_image = cv_bridge.CvBridge().imgmsg_to_cv2(msg, "bgr8")
        framesOfRGB.append(cv_image)
    except cv_bridge.CvBridgeError as e:
        rospy.logerr("CvBridge Error: %s", e)

def video_capture_callback_depth(msg):
    global recording, framesOfRGB, framesOfDepth
    if not recording:
        return
    try:
        cv_image = cv_bridge.CvBridge().imgmsg_to_cv2(msg, "mono16")
        framesOfDepth.append(cv_image)
    except cv_bridge.CvBridgeError as e:
        rospy.logerr("CvBridge Error: %s", e)


def write_video_frame_into_file(isTimeOut):
    global framesOfRGB, framesOfDepth
    is_rgb_success = False
    is_depth_success = False

    rgb_video_file_name = "rgb_video.mp4" if not isTimeOut else "rgb_video_timeout.mp4"
    depth_video_file_name = "depth_video.mp4" if not isTimeOut else "depth_video_timeout.mp4"

    if framesOfRGB:
        fourcc = cv2.VideoWriter_fourcc(*'H264')
        video_writer_rgb = cv2.VideoWriter(base_save_video_path + rgb_video_file_name, fourcc, 20.0, (framesOfRGB[0].shape[1], framesOfRGB[0].shape[0]))
        if not video_writer_rgb.isOpened():
            rospy.logerr("Failed to open rgb video writer.")
        else:
            for frame in framesOfRGB:
                video_writer_rgb.write(frame)
            video_writer_rgb.release()
            is_rgb_success = True

    if framesOfDepth:
        fourcc = cv2.VideoWriter_fourcc(*'H264')
        video_writer_depth = cv2.VideoWriter(base_save_video_path + depth_video_file_name, fourcc, 20.0, (framesOfDepth[0].shape[1], framesOfDepth[0].shape[0]))
        if not video_writer_depth.isOpened():
            rospy.logerr("Failed to open depth video writer.")
        else:
            for frame in framesOfDepth:
                video_writer_depth.write(frame)
            video_writer_depth.release()
            is_depth_success = True

    return is_rgb_success and is_depth_success


def record_video_callback(req):
    global recording, isTimeOut, framesOfRGB, framesOfDepth
    if req.record:
        recording = True
        sub_rgb = rospy.Subscriber("/usb_cam/image_raw", Image, video_capture_callback_rgb, queue_size=24)
        sub_depth = rospy.Subscriber("/camera/depth/image_raw", Image, video_capture_callback_depth, queue_size=24)

        start_time = rospy.Time.now()
        while not rospy.is_shutdown() and recording:
            if (rospy.Time.now() - start_time) > rospy.Duration(30, 0):  # 30 minutes
                rospy.loginfo("30 minutes recording limit reached, stop recording.")
                isTimeOut = True
                recording = False
                break

        success = write_video_frame_into_file(isTimeOut)
        sub_rgb.unregister()
        sub_depth.unregister()
        return RecordVideoResponse(base_save_video_path if success else "false")

    return RecordVideoResponse("false")

def stop_recording_callback(msg):
    global recording
    if msg.data:
        recording = False

if __name__ == "__main__":
    rospy.init_node("record_video_server")

    service = rospy.Service("/record_video", RecordVideo, record_video_callback)
    stop_sub = rospy.Subscriber("/record_video_stop", Bool, stop_recording_callback)

    rospy.loginfo("Ready to record videos.")

    rospy.spin()

File: src/record_video_server.cpp
#include <ros/ros.h>
#include <image_transport/image_transport.h>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include <streaming_media/RecordVideo.h>  // streaming_media 包名
#include <std_msgs/Bool.h>
#include <mutex>
#include <vector>
#include <thread>

bool recording = false, isTimeOut = false;
std::string base_save_video_path = "/home/videos/";
image_transport::Subscriber sub_rgb;
image_transport::Subscriber sub_depth;
cv::VideoWriter video_writer_rgb, video_writer_depth;
std::vector<cv::Mat> framesOfRGB, framesOfdepth;


// 视频捕获回调函数
void videoCaptureCallback(const sensor_msgs::ImageConstPtr& msg, const std::string& topic_name) {
    if (!recording) return;

    cv_bridge::CvImagePtr cv_ptr;
    try {
        if (topic_name == "rgb") {
            cv_ptr = cv_bridge::toCvCopy(msg, "bgr8");
            framesOfRGB.emplace_back(cv_ptr->image);
        } else if (topic_name == "depth") {
            cv_ptr = cv_bridge::toCvCopy(msg, "mono16");
            framesOfdepth.emplace_back(cv_ptr->image);
        }
    } catch (cv_bridge::Exception& e) {
        ROS_ERROR("cv_bridge exception: %s", e.what());
    }
}

// 写入视频文件
bool write_video_frame_into_file(bool isTimeOut) {
    bool is_rgb_success = false, is_depth_success = false;
    //std::lock_guard<std::mutex> lock(video_mutex); // 构造的时候自动锁定，析构的时候自动释放

    std::string rgbVideoFileName = isTimeOut ? "rgb_video.mp4" : "rgb_video_timeout.mp4";
    std::string depthVideoFileName = isTimeOut ? "depth_video.mp4" : "depth_video_timeout.mp4";

    if (!framesOfRGB.empty()) {
        cv::VideoWriter video_writer_rgb(base_save_video_path + rgbVideoFileName, 
                                         cv::VideoWriter::fourcc('H','2','6','4'), 20.0, 
                                         cv::Size(framesOfRGB[0].cols, framesOfRGB[0].rows));
        if (!video_writer_rgb.isOpened()) {
            ROS_ERROR("Failed to open rgb video writer.");
        } else {
            for (const auto& frame : framesOfRGB) {
                video_writer_rgb.write(frame);
            }
            video_writer_rgb.release();
            is_rgb_success = true;
            ROS_INFO("rgb video write successfully.");
        }
    }

    if (!framesOfdepth.empty()) {
        cv::VideoWriter video_writer_depth(base_save_video_path + depthVideoFileName, 
                                           cv::VideoWriter::fourcc('H','2','6','4'), 20.0, 
                                           cv::Size(framesOfdepth[0].cols, framesOfdepth[0].rows));
        if (!video_writer_depth.isOpened()) {
            ROS_ERROR("Failed to open depth video writer.");
        } else {
            for (const auto& frame : framesOfdepth) {
                video_writer_depth.write(frame);
            }
            video_writer_depth.release();
            is_depth_success = true;
        }
    }

    return is_rgb_success && is_depth_success;
}

// 服务回调函数
bool recordVideoCallback(streaming_media::RecordVideo::Request  &req, streaming_media::RecordVideo::Response &res) {
    if (req.record) {
        recording = true;
        ros::NodeHandle nh;
        image_transport::ImageTransport it(nh);
        sub_rgb = it.subscribe("/usb_cam/image_raw", 10, boost::bind(videoCaptureCallback, _1, "rgb"));
        sub_depth = it.subscribe("/camera/depth/image_raw", 10, boost::bind(videoCaptureCallback, _1, "depth"));

        ros::Time start_time = ros::Time::now();
        while (ros::ok() && recording) {
            ros::spinOnce(); // take a frame
            if ((ros::Time::now() - start_time) > ros::Duration(10.0)) { // 30 minutes
                ROS_INFO("30 minutes recording limit reached, stop recording.");
                isTimeOut = true;
                recording = false;
                break;
            }
        }


        bool success = write_video_frame_into_file(isTimeOut);
        if (success) {
            res.result = base_save_video_path;
        } else {
            res.result = "false";
        }
    }

    isTimeOut = false;
    return true;
}

void stopRecordingCallback(const std_msgs::Bool::ConstPtr& msg) {
    if (msg->data) {
        recording = false;
    }
}

int main(int argc, char** argv) {
    ros::init(argc, argv, "record_video_server");
    ros::NodeHandle nh;

    ros::ServiceServer service = nh.advertiseService("/record_video", recordVideoCallback);
    ros::Subscriber stop_sub = nh.subscribe("/record_video_stop", 1, stopRecordingCallback); // 停止录制话题

    ROS_INFO("Ready to record videos.");

    ros::AsyncSpinner spinner(2); // Use 2 threads
    spinner.start();
    ros::waitForShutdown();

    return 0;
}

